### ISSUE TITLE
All the graph have an operation _Retval that makes the train loop slow

### ISSUE THREAD

**Issue by zhaoerchao** [tensorflow/tensorflow#10730]
I have tried to train the model in [tf_cnn_benchmarks.py](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py) and [cifar10_multi_gpu_train.py](https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py).

All the timelines results like the image bellow. There is an operation named '_Retval' in the end of training loop. But as we see, long idle time was leaved in the timeline. How can I make the training faster by getting rid of the operation '_Retval' ? Is there other method to let the gpu be at full power?

![2017-06-15 19 57 31](https://user-images.githubusercontent.com/9522983/27180240-0e26c2f0-5205-11e7-9f7e-98be45a6a2b2.png)
![2017-06-15 19 58 33](https://user-images.githubusercontent.com/9522983/27180242-120350fa-5205-11e7-89f1-5d3fe6261293.png)

The images above were got by the script:

python tf_cnn_benchmarks.py local_parameter_device=cpu --num_gpus=2 --batch_size=64 --model=vgg16 --variable_update=independent --optimizer=sgd --trace_file=/home/zhaoerchao/timeline_benchmark.json --distortions

**Comment by aselle** [comment]
@zhangyaobit, could you take a look?

**Comment by zhangyaobit** [comment]
Could you mark the idle period and  _Retval  in the two images?
Could you also clarify what are the two timeline/image respectively for?  Are they for tf_cnn_benchmarks.py or cifar10_multi_gpu_train.py?

**Comment by zhaoerchao** [comment]
Thank you for your apply.

The two images in the question are all for tf_cnn_benchmarks.py

The second one is zoomed in of the first one. The _Retval operation is in the end of the image. The idle period is the period without operation that is obviously.

**Comment by zhangyaobit** [comment]
Could you upload the timeline file somewhere, so that we can take a look?

**Comment by zhaoerchao** [comment]
Sorry, I am on vacation now.  I will upload the timeline as soon as I come back.

**Comment by zhaoerchao** [comment]
[timeline_benchmark.json.txt](https://github.com/tensorflow/tensorflow/files/1100812/timeline_benchmark.json.txt)

@zhangyaobit The timeline file is uploaded, looking forward to your apply !

**Comment by zhangyaobit** [comment]
@zhaoerchao could you use a python timer to measure the total time of, say 100 steps, and then compute time per step? This will help identify if this is  an issue of Timeline or tf_cnn_benchmarks.

**Comment by zhaoerchao** [comment]
The script has measured the total time and the time per step, see [time measure code](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py#L670). I think it will have a significance performance improvement if the problem of the GPU idle time is solved.  Can you help to analysis the result of the script ?

**Comment by zhangyaobit** [comment]
What is the time per step measured by the script? If it is the time without gap, it means the gap is an artifact/bug of Timeline; if it is the time including the gap, it means the gap is performance issue of tf_cnn_benchmarks.py.

**Comment by zhaoerchao** [comment]
@zhangyaobit Thank you for your suggestion. The time per step measured including the gap. But I don't think it is a performance issue of tf_cnn_benchmarks.py. I have tried many ways to train the model but all have this issue. And [cifar10_multi_gpu_train.py](https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py) also have the issue.

Besides, when I run the script on one single GPU:
python tf_cnn_benchmarks.py local_parameter_device=cpu --num_gpus=
1 --batch_size=64 --model=resnet50 --variable_update=parameter_server --optimizer=sgd

The timeline is [timeline_benchmark_origin.json.txt](https://github.com/tensorflow/tensorflow/files/1107295/timeline_benchmark_origin.json.txt) when the script has no change and [timeline_benchmark_changed.json.txt](https://github.com/tensorflow/tensorflow/files/1107297/timeline_benchmark_changed.json.txt) when [this line](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py#L933) is replaced by 'with tf.control_dependencies([]):'.
The performance is significance improved when the operation global_step.assign_add has no dependencies.

May I ask you to clone the repository [tensorflow high performance model benchmark](https://github.com/tensorflow/benchmarks) and run it to analysis the reason ? 

Thank you very much!

**Comment by zhangyaobit** [comment]
@zheng-xq @yzhwang, could you or re-assign this to someone who is familiar with tf_cnn_benchmarks.py to take a look?

**Comment by yycho0108** [comment]
Can anyone look into this further? I'm also experiencing this issue, and it seems to me that this is either a serious issue with tensorflow's kernel execution/flow management (unlikely) or a bug of the profiler being unable to log certain operations (more probable). Either way, it is difficult to optimize a tensorflow-written pipeline when information provided from the profiling methods is opaque.

**Comment by yzhwang** [comment]
Including @prb12 since it is related to timeline. Paul could you provide some advice here? Thank you!

**Comment by prb12** [comment]
I agree that this looks strange.  The code is clearly waiting for a bunch of operations to complete via control dependencies (`group_deps_3`), and it's not clear to me from the timeline where those are executing or which one is 'slow'.

The reason for that seems to be that the device level GPU tracing (in the `/gpu:0/memcpy`  and `/gpu:0/stream:N` timelines) looks to be truncated after 12ms.  The Tensorflow GPU device is still busy queueing up CUDA kernel launches, so we know there ought to be more activity on the CUDA streams!

![image](https://user-images.githubusercontent.com/11547801/28678542-81fd0462-72a5-11e7-8c18-0264b1c24fb3.png)

There is probably a *load* of work still running on the GPU, followed by some memcpyDtoHs to transfer the results back to the host.  Eventually TF will have to wait for all of these to complete before it can update your parameters or return fetched values to the caller of session.run.

The 'group_dependencies' you commented out are (intentionally) waiting for all of that work to complete.  `_Retval` ops are the ops that return results from a TensorFlow function, and they would also typically be the point where you would need to copy results back from the device and/or sync the GPU.  (I'm not sure why this model is using functions... @zheng-xq could probably comment here)

So, to me it looks like either something is broken in the GPUTracer, or possibly the version of libcupti or the cuda drivers on your machine.

I would suggest a couple of experiments:
1) Try capturing a trace using the [NVidia tools](https://devblogs.nvidia.com/parallelforall/cuda-pro-tip-nvprof-your-handy-universal-gpu-profiler/)  -- e.g. `nvprof --print_gpu_trace` or use `nvvp` if you prefer.
2) Turn on some logging in the GPUTracer https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/default/gpu_tracer.cc

Paul

**Comment by zheng-xq** [comment]
I agree nvprof and nvvp provide very useful info.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by michaelisard** [comment]
Closing due to lack of activity, but please reopen if this still needs attention.
